lib/rex/proto/proxy/socks5.rb

```diff
@@ -0,0 +1,543 @@
+
+module Rex
+module Proto
+module Proxy
+
+#
+# A Socks5 proxy server.
+#
+class Socks5
+
+  #
+  # A client connected to the Socks5 server.
+  #
+  class Client
+
+    # COMMON HEADER FIELDS
+
+    RESERVED                        = 0
+
+    ADDRESS_TYPE_IPV4               = 1
+    ADDRESS_TYPE_DOMAINNAME         = 3
+    ADDRESS_TYPE_IPV6               = 4
+
+    # REQUEST HEADER FIELDS
+
+    REQUEST_VERSION                 = 5
+
+    REQUEST_AUTH_METHOD_COUNT               = 1
+    REQUEST_AUTH_METHOD_TYPE_NONE           = 0
+
+    REQUEST_COMMAND_CONNECT                 = 1
+    REQUEST_COMMAND_BIND                    = 2
+    REQUEST_COMMAND_UDP_ASSOCIATE           = 3     # TODO: support UDP associate
+
+    # RESPONSE HEADER FIELDS
+
+    REPLY_VERSION                      = 5
+    REPLY_FIELD_SUCCEEDED              = 0
+    REPLY_FIELD_SOCKS_SERVER_FAILURE   = 1
+    REPLY_FIELD_NOT_ALLOWED_BY_RULESET = 2
+    REPLY_FIELD_NETWORK_UNREACHABLE    = 3
+    REPLY_FIELD_HOST_UNREACHABLE       = 4
+    REPLY_FIELD_CONNECTION_REFUSED     = 5
+    REPLY_FIELD_TTL_EXPIRED            = 6
+    REPLY_FIELD_COMMAND_NOT_SUPPORTED  = 7
+    REPLY_FIELD_ADDRESS_TYPE_NOT_SUPPORTED = 8
+
+    # RPEER INDEXES
+
+    HOST                            = 1   # TODO: Unused because rpeer isn't defined.  Do we need this?
+    PORT                            = 2   # TODO: Unused because rpeer isn't defined.  Do we need this?
+
+    #
+    # A Socks5 packet.
+    #
+    class Packet
+
+      def initialize
+        @version   = REQUEST_VERSION
+        @command   = nil
+        @reserved  = RESERVED
+        @atyp      = nil
+        #@dest_port = nil   # TODO: WTF??  rpeer[HOST] isn't defined.  rpeer isn't defined.
+        #@dest_ip   = nil   # TODO: WTF?!  rpeer[HOST] isn't defined.  rpeer isn't defined.
+      end
+
+      #
+      # A helper function to recv in a Socks5 packet byte by byte.
+      #
+      # sf: we could just call raw = sock.get_once but some clients
+      #     seem to need reading this byte by byte instead.
+      #
+      def Packet.parseIncomingSessionRequest ( sock )
+        raw = ''
+        # +----+----------+----------+
+        # |VER | NMETHODS | METHODS  |
+        # +----+----------+----------+
+        # | 1  |    1     | 1 to 255 |      METHOD (\x00) = NO AUTHENTICATION REQUIRED
+        # +----+----------+----------+
+
+        version = sock.read( 1 )
+        raise "Invalid Socks5 request packet received." if not
+          ( version.unpack( 'C' ).first == REQUEST_VERSION )
+
+        nMethods = sock.read( 1 )
+        raise "Unsupported request for multiple authentication methods" if not
+          ( raw[1..1].unpack( 'C' ).first == 1 )
+
+        authMethods = sock.read( 1 )
+        raise "Unsupported request for authentication" if not
+          ( raw[2..2].unpack( 'C' ).first == 0 )
+
+        return true
+      end
+
+      def Packet.recv( sock )
+        packet = Packet.new
+
+        raw = sock.read( 262 )
```

Comment:
Just a note, I think bindata which we include as a standard library can do a lot of this unpacking and validation for you in a single step.